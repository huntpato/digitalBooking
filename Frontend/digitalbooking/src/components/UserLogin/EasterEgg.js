import { useEffect, useState } from "react";

const useInputEvent = () => {
  const [key, setKey] = useState(null);

  useEffect(() => {
    const keyDownHandler = ({ code }) => setKey(code);
    const keyUpHandler = () => setKey(null);

    global.addEventListener("keydown", keyDownHandler);
    global.addEventListener("keyup", keyUpHandler);

    return () => {
      global.removeEventListener("keydown", keyDownHandler);
      global.removeEventListener("keyup", keyUpHandler);
    };
  }, []);

  return key;
};

export const useSecretCode = (secretCode) => {
  const [count, setCount] = useState(0);
  const [success, setSuccess] = useState(false);
  const key = useInputEvent();

  useEffect(() => {
    if (key == null) return;
    if (key !== secretCode[count]) {
      setCount(0);
      return;
    }

    setCount((state) => state + 1);
    if (count + 1 === secretCode.length) {
      setSuccess(true);
    }
  }, [key]);

  return success;
};
