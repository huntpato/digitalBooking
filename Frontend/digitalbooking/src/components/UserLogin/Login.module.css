.login {
  align-items: center;
  background-color: var(--background-color2);
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  height: 93vh;
  justify-content: center;
  padding-bottom: 80px;
  padding-top: 130px;
  -webkit-animation: fade-in 0.3s cubic-bezier(0.39, 0.575, 0.565, 1) both;
  animation: fade-in 0.3s cubic-bezier(0.39, 0.575, 0.565, 1) both;
}

.login img{
  width: 400px;
}
.resLogin{
  align-items: center;
  background-color: #FAE0E0;
  border-radius: 5px;
  color:#B00020;
  display: flex;
  height: 36px;
  justify-content: center;
  align-items:center;
  margin: 0 0 15px 5px;
  padding: 0px 0px;
  width: 385px;
}

.resLogin i{
  color:#B00020;
  padding:0px;
  margin: 0px;
}

.create {
  display: flex;
  flex-direction: row;
}

.create h5 {
  color: var(--principal-color);
  font-size: 12px;
  padding: 10px 0 0 3px;
  text-decoration: none;
}

.login h2 {
  color: var(--principal-color);
  padding-bottom: 10px;
}

.item {
  padding-top: 0px;
}

.item input {
  border-radius: 5px;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.15);
  height: 25px;
  margin-left: 5px;
  padding: 15px;
  width: 385px;
}

.button {
  background-color: var(--principal-color);
  border-radius: 5px;
  border: none;
  color: var(--background-white);
  height: 30px;
  margin-top: 20px;
  transition: all 0.3s;
  width: 225px;
}

.button:hover {
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.15);
}

.button:active {
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.15);
  transform: scale(0.98);
}

.login p {
  font-size: 12px;
  padding: 10px 0px 5px 8px;
}

i {
  color: gray;
  cursor: pointer;
  margin-left: -30px;
}

.input {
  border-radius: 5px;
  border: none;
  height: 25px;
  margin-left: 5px;
  outline: none;
  padding: 15px;
  width: 250px;
}
.errorm {
  color: red;
  font-size: 12px;
  padding-top: 20px;
}

.successm {
  color: var(--principal-color);
  font-size: 12px;
  padding-top: 20px;
}


@media (max-width: 768px) {
}

@media (max-width: 480px) {
  .login input {
    height: 25px;
    width: 300px;
  }
  
  .login button {
    width: 300px;
  }

  .resLogin{
    margin: 0 0 25px 5px;
    height: 50px;
    width: 300px;
  }
}

@-webkit-keyframes fade-in {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
@keyframes fade-in {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
