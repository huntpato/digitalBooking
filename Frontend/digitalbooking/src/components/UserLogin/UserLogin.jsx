import React, { useContext } from "react";
import { useForm } from "../../hooks/useForm";
import { useState } from "react";
import { Link, useLocation } from "react-router-dom"
import { useHistory } from "react-router-dom";
import { UserContext } from "../UserContext/UserContext";
import { LoginAnimation } from "../LoginAnimation/LoginAnimation";
import { useSecretCode } from "./EasterEgg";
import parseJwt from "./parseJwt"
import styles from "./Login.module.css";

export const UserLogin = (props) => {
  const location = useLocation();
  const path= location.pathname.split("/");
  const id = path[2];
  const { setuser } = useContext(UserContext);
  const [ message, setmessage ]= useState();
  const [success,  setsuccess] = useState();
  const { login, item, input, create, button, errorm, resLogin } = styles;
  const url = "https://cors-everywhere-me.herokuapp.com/g4bookingapp-env-2.eba-4mbmxg4f.us-east-1.elasticbeanstalk.com/users/authenticate";
  const [formvalues, handleInputChange] = useForm({
    email: "",
    password: "",
  });
  const { email, password } = formvalues;
  const [showPassword, setShowPassword] = useState(false);

  const togglePassword = () => {
    setShowPassword(showPassword ? false : true);
  };
  let history = useHistory();
  const konamiCode = [
    "ArrowUp",
    "ArrowUp",
    "ArrowDown",
    "ArrowDown",
    "ArrowLeft",
    "ArrowRight",
    "ArrowLeft",
    "ArrowRight",
    "KeyB",
    "KeyA"
  ];
  const code = useSecretCode(konamiCode);
  let EasterEgg = code ? ( <img src="https://64.media.tumblr.com/tumblr_l081xsMVr51qz6z2wo1_500.gifv" alt="cosme"/>) : (<></>);
  const handleSubmit = (e) => {
    e.preventDefault();
    setmessage();
    fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(formvalues),
    })
    .then((response) => { 
      if (!response.ok) throw Error(response.status);
      return response.text()})
    .then((data) =>{
        localStorage.setItem("bearer", data);
        localStorage.setItem("data", JSON.stringify(parseJwt(data)));
        setsuccess("true");
        setmessage("Login exitoso. Redirigiendo.");
        setTimeout(() => {
           setuser(JSON.stringify(parseJwt(data)))
           history.goBack();
        }, 2000);
      })
      .catch((error) => {
        error.message == "401"? setmessage("Credenciales inválidas."):setmessage("Lamentablemente no ha podido iniciar sesión. Por favor intente más tarde.") ;
      setsuccess(false) 
    });
  };

  return (
    <form onSubmit={handleSubmit}>
      { success && <LoginAnimation message= { message }/>}
      <div className={login}>
        {id=== "resFailed" && <div className= {resLogin}><i class="fas fa-exclamation-circle"></i><p>Para realizar una reserva necesitas estar logueado</p> </div>}
        <h2>Iniciar sesión</h2>
        <div className={item}>
          <p>Correo electrónico</p>
          <input
            name="email"
            className={input}
            type="text"
            value={email}
            onChange={handleInputChange}
          />
        </div>
        <div className={item}>
          <p>Contraseña</p>
          <input
            name="password"
            className={input}
            type={showPassword ? "text" : "password"}
            value={password}
            onChange={handleInputChange}
          />
          <i
            className="fas fa-eye-slash"
            onMouseDown={togglePassword}
            onMouseUp={togglePassword}
          ></i>
        </div>
        <button name='submit' className={button}>Ingresar</button>
        <div className={create}>
          <p>¿Aun no tienes cuenta?</p>
          <Link to="/register" style={{textDecoration: 'none'}}><h5>Registrate</h5></Link>
        </div>
        {!success && <h6 className= {errorm}>{message}</h6>}
        {EasterEgg}
      </div>
    </form>
  );
};
