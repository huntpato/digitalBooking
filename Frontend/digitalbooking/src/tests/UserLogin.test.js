import React from "react";
import { shallow } from "enzyme";
import { UserLogin } from "../components/UserLogin/UserLogin";
import "@testing-library/jest-dom";
import { expect } from "@jest/globals";

jest.mock('react-router-dom', () => ({
  useParams: jest.fn().mockReturnValue({ nifUuid: 'nif123' }),
  useHistory: jest.fn(),
  useContext: jest.fn().mockReturnValue({visibility: true}),
  useLocation:jest.fn().mockReturnValue({
    pathname: "/register",
    search: "",
    hash: "",
    state: null,
    key: "5nvxpbdafa",
  }),
}));

describe("Pruebas en componente <UserLogin />", () => {
  const handleSubmit = jest.fn();
  const setuser = jest.fn();
  const wrapper = shallow(<UserLogin onSubmit={handleSubmit} />);

  test("Comprobar que <UserLogin /> renderiza correctamente.", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Verificar que el componente carga la clase de CSS correctamente", () => {
    const div = wrapper.find("div").at(0);
    expect(div.hasClass("login")).toBe(true);
  });

  test("Debe renderizar el input email", () => {
    expect(wrapper.find('input[name="email"]').exists()).toBe(true);
  });

  test("Debe renderizar el input password", () => {
    expect(wrapper.find('input[name="password"]').exists()).toBe(true);
  });

  test("Debe renderizar el boton de submit", () => {
    expect(wrapper.find('button[name="submit"]').exists()).toBe(true);
  });

  test("Verificar que los inputs se renderizan vacios", () => {
    expect(wrapper.find('input[name="email"]').prop("value")).toBe("");
    expect(wrapper.find('input[name="password"]').prop("value")).toBe("");
  });

  
});
