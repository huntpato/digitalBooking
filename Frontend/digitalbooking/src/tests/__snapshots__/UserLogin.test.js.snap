// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <UserLogin /> Comprobar que <UserLogin /> renderiza correctamente. 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="login"
  >
    <h2>
      Iniciar sesión
    </h2>
    <div
      className="item"
    >
      <p>
        Correo electrónico
      </p>
      <input
        className="input"
        name="email"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="item"
    >
      <p>
        Contraseña
      </p>
      <input
        className="input"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
      <i
        className="fas fa-eye-slash"
        onMouseDown={[Function]}
        onMouseUp={[Function]}
      />
    </div>
    <h6
      className="errorm"
    />
    <button
      className="button"
      name="submit"
    >
      Ingresar
    </button>
    <div
      className="create"
    >
      <p>
        ¿Aun no tienes cuenta?
      </p>
      <Component
        style={
          Object {
            "textDecoration": "none",
          }
        }
        to="/register"
      >
        <h5>
          Registrate
        </h5>
      </Component>
    </div>
  </div>
</form>
`;
